pkg/applicationserver/applicationserver.go

```diff
@@ -583,6 +588,208 @@ func (as *ApplicationServer) handleJoinAccept(ctx context.Context, ids ttnpb.End
+
+// recalculatePendingDownlinkQueue computes the downlink queue of the device's pending session, by decrypting the
+// invalidated queue using the device session and moving them to the pending session.
+// The previous pending session is used for cases in which the device has not been activated, and as such does not
+// have a device session. In such cases, if there are downlinks are present in the pending queue, they are decrypted
+// using the previous pending session and encrypted using new pending session.
+// This method mutates the LastAFCntDown of pending session. Downlinks which cannot be decrypted are dropped.
+// This method uses the downlink queue transaction mechanism, so any errors that occur during recomputation will
+// result in an downlink queue reset attempt.
+func (as *ApplicationServer) recalculatePendingDownlinkQueue(ctx context.Context, dev *ttnpb.EndDevice, link *link, previousPendingSession *ttnpb.Session, invalidatedDownlinks []*ttnpb.ApplicationDownlink) error {
+	return as.runDownlinkQueueTransaction(ctx, dev, link, func(ctx context.Context, dev *ttnpb.EndDevice) error {
+		var previousPendingQueue, queue []*ttnpb.ApplicationDownlink
+		unmatched := invalidatedDownlinks
+		if previousPendingSession != nil {
+			previousPendingQueue, unmatched = ttnpb.PartitionDownlinksBySessionKeyIDEquality(previousPendingSession.SessionKeyID, unmatched...)
+		}
+		if dev.Session != nil {
+			queue, unmatched = ttnpb.PartitionDownlinksBySessionKeyIDEquality(dev.Session.SessionKeyID, unmatched...)
+		}
+		logger := log.FromContext(ctx)
+		for _, item := range unmatched {
+			logger.WithFields(log.Fields(
+				"f_port", item.FPort,
+				"f_cnt", item.FCnt,
+				"session_key_id", item.SessionKeyID,
+			)).Warn("Downlink message with unknown session key ID found; dropping item")
+			registerDropDownlink(ctx, dev.EndDeviceIdentifiers, item, errUnknownSession)
+		}
+
+		var sourceQueue []*ttnpb.ApplicationDownlink
+		var sourceSession *ttnpb.Session
+		switch {
+		case previousPendingSession == nil && dev.Session == nil:
+			// We cannot decode any of the downlinks if both sessions are missing.
+			return errNoDeviceSession.New()
+		case dev.Session != nil:
+			// In the presence of a current session we copy the items from the current session to the pending session.
+			sourceQueue = queue
+			sourceSession = dev.Session
+			logger.Debug("Initializing pending downlink queue from the current session")
+		case previousPendingSession != nil:
+			// In the presence of a previous session and the absence of a current session we migrate the items from the
+			// previous pending session to the new pending session.
+			sourceQueue = previousPendingQueue
+			sourceSession = previousPendingSession
+			logger.Debug("Initializing pending downlink queue from the previous pending session")
+		}
+
+		newPendingQueue, err := as.migrateDownlinkQueue(ctx, dev.EndDeviceIdentifiers, sourceQueue, sourceSession, dev.PendingSession, 1)
```

Comment:
The queue `FCnt` cannot start from 0. It would mean that the `LastAFCntDown` is -1, which cannot the case since `LastAFCntDown` is `uint32`.

Reply:
Isn't `FCnt` starting from `0` a part of LoRaWAN certification? I know the AS part is not specified yet, but I can very much imagine that would be the requirement in the future(isn't it for NS?). Besides, there's no functional reason why `FCnt` would not be able to start from `0`, it's purely implementation-specific. You can look at NS implementation for inspiration perhaps? https://github.com/TheThingsNetwork/lorawan-stack/commit/aed9da0cfcfddd1eb43493614d8b003be282c8da As far as I understand AS does not store "recent downlinks", so it might be trickier

Reply:
> As far as I understand AS does not store "recent downlinks", so it might be trickier

Indeed, let's leave it for now

> I know the AS part is not specified yet, but I can very much imagine that would be the requirement in the future(isn't it for NS?)

AFCntDown is up to AS indeed, and that is not specified to start from 0.